### PR TITLE
Lexilla 5.1.7 + Scintilla 5.2.3

### PR DESCRIPTION
**Pull request by chcg**
Update with:
Released 22 May 2022.
    Add LexAccessor::StyleIndexAt to retrieve style values as unsigned to handle styles > 127 better. [Issue #61](https://github.com/ScintillaOrg/lexilla/issues/61).
    Associate more than one file extension with a setting for testing. [Issue #81](https://github.com/ScintillaOrg/lexilla/issues/81).
    CMake: Fix folding of "ElseIf". [Issue #77](https://github.com/ScintillaOrg/lexilla/issues/77), [Pull request #78](https://github.com/ScintillaOrg/lexilla/pull/78), [Bug #2213](https://sourceforge.net/p/scintilla/bugs/2213/).
    HTML: Fix folding of JavaScript doc comments. [Bug #2219](https://sourceforge.net/p/scintilla/bugs/2219/).
    Matlab: add "classdef" and "spmd" to folding keywords. [Pull request #70](https://github.com/ScintillaOrg/lexilla/pull/70).
    Matlab: handle "arguments" contextual keyword. [Pull request #70](https://github.com/ScintillaOrg/lexilla/pull/70).
    Matlab: improve support of class definition syntax. [Pull request #75](https://github.com/ScintillaOrg/lexilla/pull/75).
    Raku: fix escape detection. [Pull request #76](https://github.com/ScintillaOrg/lexilla/pull/76).
    Ruby: fix character sequence "?\\#" to not include '#' in SCE_RB_NUMBER as only second '\' is quoted. [Issue #69](https://github.com/ScintillaOrg/lexilla/issues/69).
    Ruby: improve styling of ternary expressions as commonly used. [Issue #69](https://github.com/ScintillaOrg/lexilla/issues/69).
    VHDL: support folding for VHDL 08 else-generate and case-generate. [Pull request #80](https://github.com/ScintillaOrg/lexilla/pull/80).

### CODE REVIEW

**Comment by chcg** [comment]
@donho See https://www.scintilla.org/ScintillaHistory.html 5.2.3:

- Duplicate APIs to support 64-bit document positions on Win32: SCI_GETTEXTRANGEFULL, SCI_FINDTEXTFULL, and SCI_FORMATRANGEFULL. This adds new types to Scintilla.iface which may impact downstream projects. Applications should move to these APIs from their predecessors as they will be deprecated.

This collides with the change done within N++ by changing the type of Sci_PositionCR. How do you think we should handle that?

**Comment by donho** [comment]
@chcg 
It's rather a good new: we can update SCI_GETTEXTRANGE, SCI_FINDTEXT, and SCI_FORMATRANGE to SCI_GETTEXTRANGEFULL, SCI_FINDTEXTFULL, and SCI_FORMATRANGEFULL respectively. OTOH, I would say to keep Notepad++'s modification `typedef intptr_t Sci_PositionCR;` though I don't see any impact if we restore it to `typedef long Sci_PositionCR;`. 

We will also encourage the plugin author update these API from their side, but it's not really urgent.

What's your option? Anyone in here has any suggestion?

**Comment by rdipardo** [comment]
@chcg, @donho, the key word in the changelog is "duplicate". N++ *already* returns 64-bit character positions from the existing APIs &#x2014; nothing has changed except the name. The old messages can become type aliases for the new ones, like this:

~~~diff
@@ -473,10 +473,10 @@
 #define SCFIND_REGEXP 0x00200000
 #define SCFIND_POSIX 0x00400000
 #define SCFIND_CXX11REGEX 0x00800000
-#define SCI_FINDTEXT 2150
 #define SCI_FINDTEXTFULL 2196
-#define SCI_FORMATRANGE 2151
+#define SCI_FINDTEXT SCI_FINDTEXTFULL
 #define SCI_FORMATRANGEFULL 2777
+#define SCI_FORMATRANGE SCI_FORMATRANGEFULL
 #define SCI_GETFIRSTVISIBLELINE 2152
 #define SCI_GETLINE 2153
 #define SCI_GETLINECOUNT 2154
@@ -488,8 +488,8 @@
 #define SCI_GETMODIFY 2159
 #define SCI_SETSEL 2160
 #define SCI_GETSELTEXT 2161
-#define SCI_GETTEXTRANGE 2162
 #define SCI_GETTEXTRANGEFULL 2039
+#define SCI_GETTEXTRANGE SCI_GETTEXTRANGEFULL
 #define SCI_HIDESELECTION 2163
 #define SCI_POINTXFROMPOSITION 2164
 #define SCI_POINTYFROMPOSITION 2165
~~~

Completely removing the old messages would create a bigger mess than #11198, since even 32-bit plugins would stop working!

**Comment by donho** [comment]
@rcdailey 
> nothing has changed except the name. 

No, it's not true. They are 2 different values in `scintilla.h` of v5.2.3:
```cpp
#define SCI_FINDTEXT 2150
#define SCI_FINDTEXTFULL 2196
```

> Completely removing the old messages would create a bigger mess than https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11198, since even 32-bit plugins would stop working!

The old messages are kept in Scintilla v5.2.3 so I see no reason to remove them from Scintilla of Notepad++.
All the plugins are compiled independently of Notepad++, they migrate these 3 API or not, they keep having always 64 bits result which is compatible with Notepad++. So the plugins won't stop working.

**Comment by rdipardo** [comment]
*Apologies to @rcdailey for the accidental mention*

---

 > They are 2 different values in scintilla.h of v5.2.3

Plugins call the message *by name*, not *by integer value*. The *name* is still `SCI_FINDTEXT`. 
I was suggesting that the *name* be used to dispatch `SCI_FINDTEXTFULL` instead; a future call to `SCI_FINDTEXT` will get a 64-bit result like it has done since 8.3, and no plugin code needs refactoring.

> The old messages are kept in Scintilla v5.2.3 so I see no reason to remove them from Scintilla of Notepad++.

According to the [mailing list announcement][0]:

> SCI_GETTEXTRANGE will be deprecated at some point (likely 6.0) and __*removed even later (7.0?)*__.

Sounds like the old messages will be around for some time yet.

[0]: https://groups.google.com/g/scintilla-interest/c/mPLwYdC0-FE#:~:text=SCI_GETTEXTRANGE%20will%20be%20deprecated%20at%20some%20point%20(likely%206.0)%20and%20removed%20even%20later%20(7.0%3F)

**Comment by rdipardo** [comment]
> I was suggesting that the name be used to dispatch `SCI_FINDTEXTFULL` instead

Of course that would be an *implementation* hack, probably touching `scintilla/src/Editor.cxx` &#x2014; not a *header* hack, as I wrongly said before. See my [edited comment](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/11734#issuecomment-1140113901).

**Comment by chcg** [comment]
Yes, with scintilla update 5.2.3 (keeping N++ change of Sci_PositionCR from long to intptr_t)
+ usage of the new interfaces

**Comment by donho** [comment]
Auto-updating of v8.4.2 has just been triggered.
This PR will be integrated in 7-10 days if there's no critical issue in v8.4.2

**Comment by donho** [comment]
@Yaron10
**Fix problem with horizontally inverted glyphs with buffered drawing and WS_EX_LAYOUTRTL set on Win32 GDI. [Feature #1435](https://sourceforge.net/p/scintilla/feature-requests/1435/).**
in https://www.scintilla.org/ScintillaHistory.html

Any idea about what it improves in Notepad++?

**Comment by Yaron10** [comment]
@donho,

It seems they're referring to https://sourceforge.net/p/csscintilla/csscintilla/ci/fa80f5d42d95cb95030f587b00174471d946353b/.
But that commit was actually merged in an earlier Scintilla version.

If I'm correct, there shouldn't be any change (or potential change) in Notepad++.

File: scintilla/win32/Scintilla.vcxproj [hunk 1]
```diff
@@ -95,6 +94,7 @@
     </ClCompile>
     <Link>
       <LinkTimeCodeGeneration>Default</LinkTimeCodeGeneration>
+      <CETCompat>true</CETCompat>
```
Comment (donho) [inline scintilla/win32/Scintilla.vcxproj:97]: Why is this line added?

**Review by donho — CHANGES_REQUESTED** [review]
Please provide the argument for adding "CETCOMPAT" tag in vsproj file.

File: scintilla/win32/Scintilla.vcxproj (hunk 1 above)
Comment (chcg) [reply scintilla/win32/Scintilla.vcxproj:97]: See Release Notes (https://www.scintilla.org/ScintillaHistory.html):
- Fix ARM64 builds with Visual C++ due to unsupported CETCOMPAT flag. [Bug #2324](https://sourceforge.net/p/scintilla/bugs/2324/).

The setting is already there at the common section, so it was just splitted out to the individual settings for x86, x64 because the feature is not available for arm64 (see also https://docs.microsoft.com/en-us/cpp/build/reference/cetcompat?view=msvc-170)

**Review by donho — APPROVED** [review]
